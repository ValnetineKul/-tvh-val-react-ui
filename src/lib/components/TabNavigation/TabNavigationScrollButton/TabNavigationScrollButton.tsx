import type { FC, MouseEventHandler } from 'react';
import React from 'react';
import IconButton from '../../Buttons/IconButton';
import Icon from '../../Icon';
import { AngleLeft, AngleRight } from '../../Icon/icons/functional';

export interface TabNavigationScrollButtonProps {
  direction: 'left' | 'right';
  disabled: boolean;
  onClick: MouseEventHandler<HTMLButtonElement>;
}

const TabNavigationScrollButton: FC<TabNavigationScrollButtonProps> = (props) => {
  const { direction, ...restProps } = props;

  if (direction === 'left') {
    return <IconButton {...restProps} size="sm" focus="inset" icon={<Icon icon={AngleLeft} />} />;
  }

  return <IconButton {...restProps} size="sm" focus="inset" icon={<Icon icon={AngleRight} />} />;
};

export default TabNavigationScrollButton;
