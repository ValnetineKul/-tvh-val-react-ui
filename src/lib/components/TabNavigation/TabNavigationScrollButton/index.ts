import TabNavigationScrollButton from './TabNavigationScrollButton';

export type { TabNavigationScrollButtonProps } from './TabNavigationScrollButton';
export default TabNavigationScrollButton;
