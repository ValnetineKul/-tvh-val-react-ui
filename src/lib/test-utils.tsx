import userEvent from '@testing-library/user-event';
import { fireEvent, render } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { theme } from '../themes/tvh';
import { createTheme, ThemeProvider, SurfaceProvider } from '../themes/core';

export const muiTheme = createTheme(theme);

const AllTheProviders: React.FC = ({ children }) => {
  return (
    <MemoryRouter>
      <ThemeProvider theme={muiTheme}>
        <SurfaceProvider color="100">{children}</SurfaceProvider>
      </ThemeProvider>
    </MemoryRouter>
  );
};

const customRender = (component: React.ReactElement, options?: Parameters<typeof render>[1]) =>
  render(component, { wrapper: AllTheProviders, ...options });

const customRenderHook: typeof renderHook = (hook, options) =>
  renderHook(hook, { wrapper: AllTheProviders, ...options });

export const initResponsiveTest = (size: number | 'Mobile' | 'Tablet' | 'Desktop' | 'LgDesktop' | 'XlDesktop') => {
  let width: number;
  if (size === 'Mobile') width = muiTheme.breakpoints.values.sm - 1;
  else if (size === 'Tablet') width = muiTheme.breakpoints.values.sm;
  else if (size === 'Desktop') width = muiTheme.breakpoints.values.md;
  else if (size === 'LgDesktop') width = muiTheme.breakpoints.values.lg;
  else if (size === 'XlDesktop') width = muiTheme.breakpoints.values.xl;
  else width = size;

  Object.defineProperty(window, 'innerWidth', {
    writable: true,
    configurable: true,
    value: width,
  });

  window.matchMedia = jest.fn().mockImplementation((query) => {
    const queryWidthInterval: string[] = query.match(/\d+(\.\d+)?/g);

    let matches;
    if (queryWidthInterval.length === 2) {
      matches = width >= +queryWidthInterval[0] && width < +queryWidthInterval[1];
    } else {
      const queryWidth = +queryWidthInterval[0];
      const isDown = /max/.test(query);

      matches = isDown ? width <= queryWidth : width >= queryWidth;
    }

    return {
      matches,
      media: query,
      onchange: null,
      addListener: jest.fn(),
      removeListener: jest.fn(),
    };
  });
};

const customUserEvent = Object.assign(userEvent, {
  mouseDown: fireEvent.mouseDown,
  mouseUp: fireEvent.mouseUp,
  mouseMove: fireEvent.mouseMove,
  scroll: fireEvent.scroll,
  error: fireEvent.error,
});

export { screen, within, waitFor, act, waitForElementToBeRemoved } from '@testing-library/react';
// override render method
export { customRender as render, customRenderHook as renderHook, customUserEvent as userEvent };
