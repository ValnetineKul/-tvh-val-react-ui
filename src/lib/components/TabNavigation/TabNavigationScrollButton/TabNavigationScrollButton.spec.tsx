import React from 'react';
import { render, screen, within } from '../../../test-utils';
import type { TabNavigationScrollButtonProps } from './TabNavigationScrollButton';
import TabNavigationScrollButton from './TabNavigationScrollButton';

describe('TabNavigationScrollButton', () => {
  it('Should render correctly', () => {
    const { container } = render(<TabNavigationScrollButton onClick={jest.fn()} direction="left" disabled={false} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  describe('Should render correct icon inside scroll button', () => {
    const cases = [
      ['left', 'angleLeft.svg'],
      ['right', 'angleRight.svg'],
    ];
    test.each(cases)('given %p, returns %p', (input: TabNavigationScrollButtonProps['direction'], expected: string) => {
      render(<TabNavigationScrollButton onClick={jest.fn()} direction={input} disabled={false} />);
      const scrollButton = screen.getByRole('button');
      expect(within(scrollButton).getByText(expected, { selector: 'svg' })).toBeInTheDocument();
    });
  });

  it('Should trigger onClick', () => {
    const handleClick = jest.fn();
    render(<TabNavigationScrollButton onClick={handleClick} direction="left" disabled={false} />);
    const scrollButton = screen.getByRole('button');
    scrollButton.click();
    expect(handleClick).toHaveBeenCalled();
  });

  it('Should disable button if disabled prop is true', () => {
    render(<TabNavigationScrollButton onClick={jest.fn()} direction="left" disabled />);
    const scrollButton = screen.getByRole('button');
    scrollButton.click();
    expect(scrollButton).toBeDisabled();
  });
});
