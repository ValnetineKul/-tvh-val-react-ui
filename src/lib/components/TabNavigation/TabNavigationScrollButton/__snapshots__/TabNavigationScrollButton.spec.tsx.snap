// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabNavigationScrollButton Should render correctly 1`] = `
<button
  class="MuiButtonBase-root tss-mlornr-ButtonBase-root-ButtonBase-focusInset-IconButton-root mui-10d1a0h-MuiButtonBase-root"
  tabindex="0"
  type="button"
>
  <div
    class="tss-4fpoey-Icon-root-Icon-sizeSm"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium mui-i4bv87-MuiSvgIcon-root"
      color="inherit"
      focusable="false"
      viewBox="0 0 24 24"
    >
      angleLeft.svg
    </svg>
  </div>
</button>
`;
